### PR TITLE
Fixing issues around domain-less Constants and adjoint

### PR DESCRIPTION
**Pull request by stephankramer**
Changes required to fix adjoint:
* Constants used as control (and any Constant that changes value during the run) requires a domain for the adjoint to work
* callbacks pre and post derivative calculation of a ReducedFunctional now need to pass through resp. the control and the derivatives
* avoid assigning to Constants internally as this causes issues with the adjoint - notably this was done in updating fields_old in some timesteppers (we maybe should just get rid of fields_old altogether)

### CODE REVIEW

File: thetis/timeintegrator.py [hunk 1]
```diff
@@ -85,6 +85,32 @@ def advance_picard(self, t, update_forcings=None, update_lagged=True, update_fie
...
+                    # are now just Functions on the Real space, so are dealt with under isinstance(..., Function)
+                    self.fields_old[k] = self.fields[k]
+
+    def update_fields_old(self):
+        """
+        Update the values of fields_old with those of fields
+        """
+        for k in sorted(self.fields):
+            if isinstance(self.fields[k], Function):
+                self.fields_old[k].assign(self.fields[k])
```
Comment (jwallwork23) [inline thetis/timeintegrator.py:112]: Looks good. It's also a nice refactoring to move this to the base class anyway :)

File: examples/discrete_turbines/channel-optimisation.py [hunk 2]
```diff
@@ -166,7 +166,8 @@ def eval_cb_pre(controls):
 
 def derivative_cb_pre(controls):
     print_output("ADJOINT RUN:")
-    print_output("positions: {}".format([float(c_) for c_ in controls]))
+    print_output("positions: {}".format([float(c_.data()) for c_ in controls]))
```
Comment (jwallwork23) [inline examples/discrete_turbines/channel-optimisation.py:169]: Out of interest, why was this change needed? Does taking the `float` of a `Constant` no longer do what you expect?

**Review by jwallwork23 — APPROVED** [review]
Thanks Stephan, this all looks great. Good typo catches, too.

File: examples/discrete_turbines/channel-optimisation.py (hunk 2 above)
Comment (stephankramer) [reply examples/discrete_turbines/channel-optimisation.py:169]: Ah, because if you create a `Constant(..., domain=...)` you no longer actually get an object of type `Constant` it actually returns a `Function` on the R FunctionSpace. And taking the float of a Function is not defined of course.

Comment (jwallwork23) [reply examples/discrete_turbines/channel-optimisation.py:169]: Ah okay makes sense, thanks.
